## import the required libraries.

In [1]:
import configparser
import pandas as pd
import numpy as np
import requests
import tweepy
import json
import os

# Data Gathering

## Load the `twitter archive enhanced csv` file into the pandas dataframe

In [2]:
Dog_tweets_df = pd.read_csv('Datasets/twitter-archive-enhanced.csv')

## Download the tweet image predictions

In [3]:
# # Check the if the required folder exists. If it does not exist, create it
# folder_name = 'Datasets'
# if not os.path.exists (folder_name):
#     os.makedirs(folder_name)

# # Use the request.get method to get access the data from the server.
# url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
# response = requests.get(url)

# # Save the responses in a .tsv file
# with open (os.path.join(folder_name,url.split('/')[-1]), mode = 'wb') as file:
#     file.write(response.content)

### Load the Image prediction file into the pandas dataframe

In [4]:
image_prediction_df = pd.read_csv('Datasets/image-predictions.tsv', sep='\t')

## Get additional data from the Twitter API

In [5]:
# Read the configurations
config = configparser.ConfigParser()
config.read('config.ini')

api_key = config['twitter']['api_key']
api_key_secret = config['twitter']['api_key_secret']

access_token = config['twitter']['access_token']
access_token_secret = config['twitter']['access_token_secret']

In [6]:
# Authentication
auth = tweepy.OAuthHandler(api_key, api_key_secret)
auth.set_access_token(access_token, access_token_secret)

api = tweepy.API(auth, wait_on_rate_limit=True)

In [7]:
# # Extract all the tweet IDs in the WeRateDogs Twitter archive and save them in a python list.
# tweet_response_counts = []
# for tweet_id in Dog_tweets_df.tweet_id:
#     try:
#         tweet_info = api.get_status(tweet_id, tweet_mode='extended')
#     except Exception:
#         pass
#     finally:
#         tweet_id = tweet_info.id
#         retweet_count = tweet_info.retweet_count
#         favorite_count = tweet_info.favorite_count
        
#         tweet_response_counts.append({'tweet_id':tweet_id,
#                                       'retweet_count':retweet_count,
#                                       'favorite_count':favorite_count})
#         print(tweet_id)
# counts_df = pd.DataFrame(tweet_response_counts, columns = ['tweet_id', 'retweet_count', 'favorite_count'])

In [8]:
# count = 0
# fails_dict = {}
# start = timer()
# # Save each tweet's returned JSON as a new line in a .txt file
# with open('tweet_json.txt', 'w') as outfile:
#     # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
#     for tweet_id in tweet_ids:
#         count += 1
#         print(str(count) + ": " + str(tweet_id))
#         try:
#             tweet = api.get_status(tweet_id, tweet_mode='extended')
#             print("Success")
#             json.dump(tweet._json, outfile)
#             outfile.write('\n')
#         except tweepy.TweepError as e:
#             print("Fail")
#             fails_dict[tweet_id] = e
#             pass
# end = timer()
# print(end - start)
# print(fails_dict)

In [9]:
# counts_df.to_csv('Datasets/reaction_counts.csv', index=False)

### Load the dataframe into the pandas dataframe

In [10]:
reaction_counts_df = pd.read_csv('Datasets/reaction_counts.csv')

# Assessing Data

## 1. Assessing the Dog_tweet dataframe

In [11]:
# Display the dataframe
Dog_tweets_df

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

In [12]:
Dog_tweets_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [13]:
# Check whether there are abnormal values in the rating numerator column
Dog_tweets_df.rating_numerator.describe()

count    2356.000000
mean       13.126486
std        45.876648
min         0.000000
25%        10.000000
50%        11.000000
75%        12.000000
max      1776.000000
Name: rating_numerator, dtype: float64

> Note that the maximum numerator rating is abnormally high. Let use view that particular record to gain more insight about it.

In [14]:
Dog_tweets_df.query('rating_numerator == 1776.000000')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
979  749981277374128128                    NaN                  NaN   

                     timestamp  \
979  2016-07-04 15:00:45 +0000   

                                                source  \
979  <a href="https://about.twitter.com/products/tw...   

                                                  text  retweeted_status_id  \
979  This is Atticus. He's quite simply America af....                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
979                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
979  https://twitter.com/dog_rates/status/749981277...              1776   

     rating_denominator     name doggo floofer pupper puppo  
979                  10  Atticus  None    None   None  None

> It appears that the entry is not a problem. Perhaps, the person who rated the dog liked it that much.

In [15]:
# Check whether there are abnormal values in the rating denominator column
Dog_tweets_df.rating_denominator.describe()

count    2356.000000
mean       10.455433
std         6.745237
min         0.000000
25%        10.000000
50%        10.000000
75%        10.000000
max       170.000000
Name: rating_denominator, dtype: float64

In [16]:
Dog_tweets_df.duplicated().sum()

0

In [17]:
Dog_tweets_df.isnull().sum()

tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

## 2. Assessing the Image Predictions Dataframe

In [18]:
# Display the dataframe
image_prediction_df

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

In [19]:
# Check the datatypes of every column an whether any column has null entries
image_prediction_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [20]:
# Check for duplicates
image_prediction_df.duplicated().sum()

0

In [21]:
# Check if the dataframe has abnormal numerical values.
image_prediction_df.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

## 3. Assessing the `likes and retweet` dataframe

In [22]:
# Display the dataframe
reaction_counts_df

tweet_id  retweet_count  favorite_count
0     892420643555336193           7009           33815
1     892177421306343426           5302           29320
2     891815181378084864           3481           22050
3     891689557279858688           7219           36903
4     891327558926688256           7762           35311
...                  ...            ...             ...
2351  666049248165822465             37              89
2352  666044226329800704            115             247
2353  666033412701032449             36             100
2354  666029285002620928             39             112
2355  666020888022790149            423            2293

[2356 rows x 3 columns]

In [23]:
# Check for null entries in the dataframe and the datatype of every column.
reaction_counts_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2356 non-null   int64
 1   retweet_count   2356 non-null   int64
 2   favorite_count  2356 non-null   int64
dtypes: int64(3)
memory usage: 55.3 KB


In [24]:
# Check for abnormal numerical values.
reaction_counts_df.describe()

tweet_id  retweet_count  favorite_count
count  2.356000e+03    2356.000000      2356.00000
mean   7.427737e+17    2492.997878      7157.22708
std    6.857010e+16    4193.078114     11028.70282
min    6.660209e+17       1.000000         0.00000
25%    6.783989e+17     500.750000      1243.00000
50%    7.196279e+17    1186.000000      3116.00000
75%    7.993373e+17    2879.000000      8857.50000
max    8.924206e+17   70758.000000    144922.00000

# Assessment Report

### Quality Issues
#### Dog_tweet dataframe
1. Missing tweet information in the following columns:
> - in_reply_to_status_id (2278 null entries)<br>
> - in_reply_to_user_id (2278 null entries)<br>
> - retweeted_status_id (2175 null entries)<br>
> - retweeted_status_user_id (2175 null entries)<br>
> - retweeted_status_timestamp (2175 null entries)<br>
> - expanded_urls (59 null entries)<br>
2. Erroneous datatypes (timestamp, retweeted_status_timestamp columns, rating_denominator, rating_numerator columns).
3. `source column` contain a tags (<a/>).
4. Some of the entries in the `expanded_urls` column contains double entries that are duplicates.
#### image prediction dataframe
1. This dataframe has non-descriptive column names.
2. Comparing the number of entries in the image prediction dataframe and the dog tweets dataframe, it is clear that some of the tweets have no corresponding image predictions in the image prediction dataframe.
#### Favourite counts & retweet counts dataframe
- No issues detected

### Tidiness Issues
#### Dog_tweet dataframe
1. `timestamp column`  has two variable entries (Date & Time).
2. `retweeted_status_timestamp` has two variable entries (Date & Time).
2. `source column` has two variables (source name & source url).
3. `text column` has four variables in the same column (The tweet, rating_numerator, rating_denominator, Tweet link). **note: the tweet link in this column is similar to the link in the source column**
4. The `doggo`, `floofer`, `pupper`, `puppo` should be variables.
#### image prediction dataframe
- No isses detected
#### Favourite counts & retweet counts dataframe
1. We do not need this table because the information it holds (favourite counts and retweet counts) should appear alongside each respective tweet in the dog tweet dataframe.
